substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -672,11 +674,28 @@ pub mod pallet {
 
             log::info!(
                 "{:?} contracts to bill at block {:?}",
-                contracts,
+                contract_ids,
                 block_number
             );
 
-            for contract_id in contracts {
+            for contract_id in contract_ids {
+                if let Some(c) = Contracts::<T>::get(contract_id) {
+                    if let types::ContractData::NodeContract(node_contract) = c.contract_type {
+                        let bill_ip = node_contract.public_ips > 0;
+                        let bill_cu_su = NodeContractResources::<T>::contains_key(contract_id)
+                            && !NodeContractResources::<T>::get(contract_id).used.is_empty();
```

Comment:
why `contains_key` and `get` afterwards ? I think `get` suffices because of the `ValueQuery` storage definition (default when not present)

Reply:
Indeed
But I have a conceptual case here
Ok, I am aware the `get`  will return default value
And what if a default value was intentionally pushed? how to differentiate both scenarios? (nothing pushed / default value pushed)

Reply:
> And what if a default value was intentionally pushed?

This is something that is bad practice and should be avoided, it also results in the same. To me, this looks like 2 storage reads while it can be only 1 storage read.

Reply:
Exactly, in our case it results the same
But if we change the default value for the resource we are broken
So, I ll get rid of extra storage read and add comment

Reply:
Why are we broken if we change the default value for resources? `.used.is_empty()` should always check if all values are 0 and this should not change

Reply:
if for some reason we set this for example

~~~rust
    pub fn default() -> Resources {
        Resources {
            hru: 1,
            sru: 1,
            cru: 1,
            mru: 1,
        }
    }

    pub fn empty() -> Resources {
        Resources {
            hru: 0,
            sru: 0,
            cru: 0,
            mru: 0,
        }
    }

Reply:
well, if the default is any other than 0 values, the user has to pay anyway